org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/SbtScopesBuildManager.scala

```diff
@@ -0,0 +1,85 @@
+package org.scalaide.core.internal.project
+
+import scala.tools.nsc.Settings
+
+import org.eclipse.core.resources.IFile
+import org.eclipse.core.resources.IProject
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.runtime.IProgressMonitor
+import org.eclipse.core.runtime.SubMonitor
+import org.scalaide.core.IScalaPlugin
+import org.scalaide.core.IScalaProject
+import org.scalaide.core.SdtConstants
+import org.scalaide.core.internal.builder.BuildProblemMarker
+import org.scalaide.core.internal.builder.EclipseBuildManager
+import org.scalaide.core.internal.builder.zinc.SbtAwareBuildManager
+import org.scalaide.core.internal.project.scopes.BuildScopeUnit
+import org.scalaide.ui.internal.preferences.ScalaPluginSettings
+import org.scalaide.util.internal.SettingConverterUtil
+
+import sbt.inc.Analysis
+import sbt.inc.IncOptions
+
+trait ScopeBuildManager extends EclipseBuildManager {
+  def hasErrors(scope: CompileScope): Boolean
+}
+
+class SbtScopesBuildManager(val owningProject: IScalaProject, managerSettings: Settings)
+    extends SbtAwareBuildManager {
+  val DefaultScopesOrder: Seq[CompileScope] = Seq(CompileMacrosScope, CompileMainScope, CompileTestsScope)
+
+  private val buildScopeUnits = DefaultScopesOrder.foldLeft(List.empty[BuildScopeUnit]) { (acc, scope) =>
+    new BuildScopeUnit(scope, owningProject, managerSettings, acc.toSeq) :: acc
+  }.reverse
+
+  def hasErrors(compileScope: CompileScope): Boolean =
+    buildScopeUnits filter { _.scope == compileScope } exists { _.hasErrors }
+
+  override def hasErrors = hasInternalErrors && (buildScopeUnits exists { _.hasErrors })
+
+  def build(addedOrUpdated: Set[IFile], removed: Set[IFile], monitor: SubMonitor): Unit = {
+    owningProject.underlying.deleteMarkers(SdtConstants.ProblemMarkerId, true, IResource.DEPTH_INFINITE)
+    hasInternalErrors = buildScopeUnits map { buildScopeUnit =>
+      if (areTransitiveDependenciesBuilt(buildScopeUnit)) {
+        buildScopeUnit.sources.foreach {
+          _.deleteMarkers(SdtConstants.ProblemMarkerId, true, IResource.DEPTH_INFINITE)
```

Comment:
Maybe the original code was duplicating work as well, entirely possible :) 

So, line 41 deletes all markers in the current project. What is there left to be deleted in line 45? Aren't all the sources in each build scope "under" the enclosing project?